src/main/java/picard/analysis/WgsMetricsProcessorImpl.java

```diff
@@ -0,0 +1,126 @@
+ * THE SOFTWARE.
+ */
+
+package picard.analysis;
+
+import htsjdk.samtools.metrics.MetricsFile;
+import htsjdk.samtools.reference.ReferenceSequence;
+import htsjdk.samtools.reference.ReferenceSequenceFileWalker;
+import htsjdk.samtools.util.*;
+import picard.PicardException;
+import picard.filter.CountingFilter;
+import picard.filter.CountingPairedFilter;
+
+import java.util.Arrays;
+import java.util.stream.LongStream;
+
+/**
+ * Implementation of <code>WgsMetricsProcessor</code> that gets input data from a given iterator
+ * and processes it with a help of collector
+ * @author Mariia_Zueva@epam.com, EPAM Systems, Inc. <www.epam.com>
+ */
+public class WgsMetricsProcessorImpl<T extends AbstractRecordAndOffset> implements WgsMetricsProcessor {
+    /**
+     * Source of input data
+     */
+    private final AbstractLocusIterator<T, AbstractLocusInfo<T>> iterator;
+    /**
+     * Accumulates the data from iterator
+     */
+    private final AbstractWgsMetricsCollector<T> collector;
+    /**
+     * ReferenceWalker for a processed reference sequence
+     */
+    private final ReferenceSequenceFileWalker refWalker;
+    /**
+     * Logger for the progress of work
+     */
+    private final ProgressLogger progress;
+
+    private final Log log = Log.getInstance(WgsMetricsProcessorImpl.class);
+
+    /**
+     * @param iterator  input <code>AbstractLocusIterator</code>
+     * @param refWalker over processed reference file
+     * @param collector input <code>AbstractWgsMetricsCollector</code>
+     * @param progress  logger
+     */
+    public WgsMetricsProcessorImpl(AbstractLocusIterator<T, AbstractLocusInfo<T>> iterator,
+            ReferenceSequenceFileWalker refWalker,
+            AbstractWgsMetricsCollector<T> collector,
+            ProgressLogger progress) {
+        this.iterator = iterator;
+        this.collector = collector;
+        this.refWalker = refWalker;
+        this.progress = progress;
+    }
+
+    /**
+     * Method gets the data from iterator for each locus and processes it with the help of collector.
+     */
+    @Override
+    public void processFile() {
+        long counter = 0;
+
+        while (iterator.hasNext()) {
+            final AbstractLocusInfo<T> info = iterator.next();
+            final ReferenceSequence ref = refWalker.get(info.getSequenceIndex());
+            boolean referenceBaseN = collector.isReferenceBaseN(info.getPosition(), ref);
+            collector.addInfo(info, ref, referenceBaseN);
+            if (referenceBaseN) {
+                continue;
+            }
+
+            progress.record(info.getSequenceName(), info.getPosition());
+            if (collector.isTimeToStop(++counter)) {
+                break;
+            }
+            collector.setCounter(counter);
+        }
+        // check that we added the same number of bases to the raw coverage histogram and the base quality histograms
+        final long sumBaseQ= Arrays.stream(collector.unfilteredBaseQHistogramArray).sum();
```

Comment:
This code in the master branch was moved out of the cycle, so that check is performed only once after the whole metric's run. We've made the same changes here.